client-apis/encrypt.md

```diff
@@ -130,72 +180,180 @@ This document refers to the output of the key derivation algorithm as the derive
+  outputted when calculating the encrypted content above.
+
+Once the entire frame is serialized,
+the serialized frame MAY be released.
+
+If the algorithm suite contains a signature algorithm and
+this operation is [streaming](streaming.md) the encrypted message output to the caller,
+this operation MUST input the serialized frame to the signature algorithm as soon as it is serialized,
+such that the serialized frame isn't required to remain in memory to complete
+the [signature calculation](#signature-calculation).
+
+### Construct the signature
+
 If the [algorithm suite](../framework/algorithm-suites.md) contains a [signature algorithm](../framework/algorithm-suites.md#signature-algorithm),
-the output [message](../data-format/message.md) MUST contain a [message footer](../data-format/message-footer.md).
-The footer is serialized with the following specifics:
+this operation MUST calculate a signature over the message,
+and the output [encrypted message](#encrypted-message) MUST contain a [message footer](../data-format/message-footer.md).
+
+To calculate a signature, this operation MUST use the [signature algorithm](../framework/algorithm-suites.md#signature-algorithm)
+specified by the [algorithm suite](../framework/algorithm-suites.md), with the following input:
+
+- the signature key is the [signing key](../framework/structures.md#signing-key) in the [encryption materials](../framework/structures.md#encryption-materials)
+- the input to sign is the concatenation of the serialization of the [message header](../data-format/message-header.md) and [message body](../data-format/message-body.md)
+
+Note that the message header and message body MAY have already been inputted during previous steps.
 
-- [Signature](../data-format/message-footer.md#signature): MUST be the output of the [signature algorithm](../framework/algorithm-suites.md#signature-algorithm)
-  specified by the [algorithm suite](../framework/algorithm-suites.md), with the following input:
-  - the signature key is the [signing key](../framework/structures.md#signing-key) in the [encryption materials](../framework/structures.md#encryption-materials)
-  - the input to sign is the concatenation of the serialization of the [message header](../data-format/message-header.md) and [message body](../data-format/message-body.md)
+This operation MUST then serialize a message footer with the following specifics:
 
-Any data that is not specified within the [message format](../data-format/message.md) MUST NOT be added to the output message.
+- [Signature Length](../data-format/message-footer.md#signature-length): MUST be the length of the
+  output of the calculation above.
+- [Signature](../data-format/message-footer.md#signature): MUST be the output of the calculation above.
+
+Once the entire message is signed and the message footer is serialized,
+the serialized message footer MAY be released.
```

Comment:
Same as above, although in this case it feels like this statement and the next one could just be combined into one, since there's only one correct moment in time to release this.